### DISCUSSION TITLE
C# Nullable analyzer shouldn't warn on unused method parameters

### DISCUSSION THREAD

**Discussion by SetTrend** [dotnet/csharplang#4352] [General]
This piece of code shouldn't yield a warning message:

```c#
public bool CanExecute(object _) => true;
```

The method parameter isn't used here, so I suggest no warning to be emitted.

Currently, the following warning is generated:

```
Severity   Code     Description
Warning    CS8767   Nullability of reference types in type of parameter '_' of 'bool AddCommand.CanExecute(object _)' doesn't match implicitly implemented member 'bool ICommand.CanExecute(object? parameter)' (possibly because of nullability attributes).
```

**Comment by huoyaoyuan** [comment]
The signature checking part doesn't consider method body at all. Doing so will break the modular design of compiler.

**Reply by jnm2** [reply]
It could work by ignoring parameters named `_`, since you'll be using names like that to avoid another unused parameter warning.

**Comment by jnm2** [comment]
I think it's just as important to not have a misleading signature even when parameters are not used.

**Reply by SetTrend** [reply]
What for?

**Reply by 333fred** [reply]
That method has to handle null. Today, it handles it by never touching the parameter, but there's no guarantee that it'll still be doing that tomorrow. It would be very odd if, during a refactoring, adding code to the body of a method suddenly introduced new errors about the signature of the method.

**Reply by SetTrend** [reply]
To me, it would actually make sense if such error arose only then.

I wouldn't expect, for example, to get a warning for `int` values exceeding a value of `255`, just because I might, sometime in the future, refactor a function and change the data type to `byte`.

**Reply by 333fred** [reply]
> I wouldn't expect, for example, to get a warning for int values exceeding a value of 255, just because I might, sometime in the future, refactor a function and change the data type to byte.

Sure, but that's not what we're talking about here. The equivalent would be if you were overriding a method that accepted an `int` with a method that accepted a `byte`. That's just not legal: you're being passed an int, not a byte. It doesn't matter that you don't _use_ that parameter: what the function accepts is an int, not a byte. Similarly, the method you're overriding does not accept `object`, it accepts `object?`. It is not safe to convert that, and so you get told about it.

**Reply by SetTrend** [reply]
Ah, OK.

**Reply by SetTrend** [reply]
So, if the reason for this warning is merely based on a _**mismatch**_ between declaration (in a base class or interface) and its implementation, then that perfectly makes sense.

Thank you for taking the time.
